clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -2070,4 +2086,318 @@ void handleFailure(Throwable throwable) {
+        final ArrayList<String> groupIdList = new ArrayList<>();
+        for (String groupId : groupIds) {
+            if (!consumerGroupFutures.containsKey(groupId)) {
+                consumerGroupFutures.put(groupId, new KafkaFutureImpl<ConsumerGroupDescription>());
+                groupIdList.add(groupId);
+            }
+        }
+
+        for (final String groupId : groupIdList) {
+
+            final long nowFindCoordinator = time.milliseconds();
+            final long deadline = calcDeadlineMs(nowFindCoordinator, options.timeoutMs());
+
+            runnable.call(new Call("findCoordinator", deadline, new ControllerNodeProvider()) {
+                @Override
+                AbstractRequest.Builder createRequest(int timeoutMs) {
+                    return new FindCoordinatorRequest.Builder(FindCoordinatorRequest.CoordinatorType.GROUP, groupId);
+                }
+
+                @Override
+                void handleResponse(AbstractResponse abstractResponse) {
+                    final FindCoordinatorResponse response = (FindCoordinatorResponse) abstractResponse;
+
+                    final long nowDescribeConsumerGroups = time.milliseconds();
+
+                    final int nodeId = response.node().id();
+
+                    runnable.call(new Call("describeConsumerGroups", deadline, new ConstantNodeIdProvider(nodeId)) {
+
+                        @Override
+                        AbstractRequest.Builder createRequest(int timeoutMs) {
+                            return new DescribeGroupsRequest.Builder(groupIdList);
+                        }
+
+                        @Override
+                        void handleResponse(AbstractResponse abstractResponse) {
+                            final DescribeGroupsResponse response = (DescribeGroupsResponse) abstractResponse;
+                            // Handle server responses for particular groupId.
+                            for (Map.Entry<String, KafkaFutureImpl<ConsumerGroupDescription>> entry : consumerGroupFutures.entrySet()) {
+                                final String groupId = entry.getKey();
+                                final KafkaFutureImpl<ConsumerGroupDescription> future = entry.getValue();
+                                final DescribeGroupsResponse.GroupMetadata groupMetadata = response.groups().get(groupId);
+                                final Errors groupError = groupMetadata.error();
+                                if (groupError != Errors.NONE) {
+                                    future.completeExceptionally(groupError.exception());
+                                    continue;
+                                }
+
+                                final String protocolType = groupMetadata.protocolType();
+                                if (protocolType.equals(ConsumerProtocol.PROTOCOL_TYPE) || protocolType.isEmpty()) {
```

Comment:
This is a meta suggestion, not a comment for the changes of this PR: although currently in AK we only have two protocol types, it is possible that we will add more in AK (e.g. Kafka Streams may implement its own coordiantor), and also non-AK code can customize their own coordinator implementation with different protocol names. So it may be less appealing to add one API per each protocol type with such filtering in this line as it is 1) less extensible and 2) not usable for other non-builtin protocol types.

I'd suggest: 

a. In this PR we keep the `XXXConsumerGroups` APIs in admin client, and complete this as KIP-222.
b. In the future we can add another set of `XXXGroups(String protocolType, ..)` API.

The hope is that a) is commonly used for group queries and hence worth making it a special "syntax sugar" for most users, while only advanced users would like to use b) for other protocol types.

WDYT @jeqo @hachikuji ?

Reply:
@guozhangwang agree. Actually I was wondering about `deleteConsumerGroups` as in the API there is no way (I'm aware of) to validate that we are deletign a *ConsumerGroup* and not another type of Group (e.g. Connect). Maybe this is a case to add `deleteGroups` instead of the current one?